src/fastpath/fastpath.c

```diff
@@ -6,15 +6,254 @@
+                if (nativeThreadUsingFPU(dest)) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+#endif
+#endif
+            } else {
+                sc = dest->tcbSchedContext;
+            }
+
+            /* Only fastpath signal to threads which will not become the new highest prio thread on the
+             * core of their SC, even if the currently running thread on the core is the idle thread. */
+            if (NODE_STATE_ON_CORE(ksCurThread, sc->scCore)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
+
+            /* Simplified schedContext_resume that does not change state and reverts to the
+            * slowpath in cases where the SC does not have sufficient budget, as this case
+            * adds extra scheduler logic. Normally, this is done after donation of SC
+            * but after tweaking it, I don't see anything executed in schedContext_donate
+            * that will affect the conditions of this check */
+            if (sc->scRefillMax > 0) {
+                if (!(refill_ready(sc) && refill_sufficient(sc, 0))) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+                schedulable = true;
+            }
+
+            /* Check if signal is cross-core or cross-domain */
+            if (ksCurDomain != dest->tcbDomain SMP_COND_STATEMENT( || sc->scCore != getCurrentCPUIndex())) {
+                crossnode = true;
+            }
+
+            /*  Point of no return */
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+            /* Cancel the IPC that the signalled thread is waiting on */
+            cancelIPC_fp(dest);
+
+            /* Unblock up the signalled thread and tranfer badge */
+            setRegister(dest, badgeRegister, badge);
+            thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+
+            /* Donate SC if necessary. The checks for this were already done before
+            * the point of no return */
+            maybeDonateSchedContext_fp(dest, sc);
+
+            /* Left this in the same form as the slowpath. Not sure if optimal */
+            if (sc_sporadic(dest->tcbSchedContext)) {
+                assert(dest->tcbSchedContext != NODE_STATE(ksCurSC));
+                if (dest->tcbSchedContext != NODE_STATE(ksCurSC)) {
+                    refill_unblock_check(dest->tcbSchedContext);
+                }
+            }
+
+            /* If dest was already not schedulable prior to the budget check
+             * the slowpath doesn't do anything special besides not
+             * not scheduling the dest thread. */
+            if (schedulable) {
+                if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority || crossnode) {
+                    SCHED_ENQUEUE(dest);
+                } else {
+                    SCHED_APPEND(dest);
+                }
```

Comment:
`NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority` is always false for the current node, as this is checked on line 96. The corner case is caught by `crossnode` already.

```suggestion
                if (crossnode) {
                    SCHED_ENQUEUE(dest);
                } else {
                    SCHED_APPEND(dest);
                }
```

Reply:
I think this condition should actually be the other way around ->  `NODE_STATE(ksCurThread)->tcbPriority > dest->tcbPriority` to match with slowpath behavior as can be seen [here](https://github.com/seL4/seL4/blob/3978092885e4f3e6524fb3e40b04c02b35804c50/src/kernel/thread.c#L381). This mistake must have been an artifact of our previous discussion on the correctness of append/enqueue.

Reply:
Agreed, you're probably right. The corner case checked here is whether the `dest` priority is equal or lower.